Project-Management/CoreTeam/workspace-notes.md

```diff
@@ -0,0 +1,43 @@
+# Workspace
+
+## Issues with the current workspace
+
+Focussing mainly on `MatrixWorkspace`.
+
+1. Constructed iteratively, i.e., workspaces start their life in an incomplete and dysfunctional state, vital information is added only later (size, instrument, spectrum to detector mapping).
+1. No encapsulation for data it contains, thus violating invariants:
+   - Access to X, Y, and E makes breaking or abusing internal data easy.
+   - Access to spectrum numbers on `ISpectrum` prevents validation (e.g., duplicates).
+   - Detector IDs can be set arbitrarily on `ISpectrum`, no check if this is in the instrument.
+1. `MatrixWorkspace` is meant to store histograms, but is used for many other purposes:
+   - Single values.
+   - Single data points in each histogram. Leads to performance problems due to massive overhead from using histograms for this purpose, e.g., at ILL.
+   - Masking.
+   - Grouping.
+   - HKL values.
+1. Not enough workspaces types (thus the aforementioned misappropriation of `MatrixWorkspace`).
+   - In the current way workspaces and algorithms are built, adding new workspaces types is not very useful since existing algorithms will not work for them.
+1. The base class of `MatrixWorkspace`, `ExperimentInfo` is a complicated and holds a lot of complexity related to setting the instrument and loading parameter maps.
```

Comment:
I think the whole virtual inheritance approach should be reconsidered. There is not much that a base `Workspace` actually gives you, and it has lead to the widespread usage of down-casting (which you have already mentioned). Furthermore, I think we should drop the concept that a `Workspace` subtypes are a pluggable thing. They really aren't (at least in their current form) and that has lead to a lot of shoe-horning, maybe explaining the bloated interfaces you have already mentioned.

Reply:
Performance is another reason to be careful with virtual inheritance. Changes to vtkDataSet to avoid virtual functions in tight loops improved performance by a factor of 2-3x.

[2017-03-18-Highlights.pdf](https://github.com/mantidproject/documents/files/1383875/2017-03-18-Highlights.pdf)